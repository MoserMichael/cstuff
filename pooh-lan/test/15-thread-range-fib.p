# co-routine threads: create thread via API, thread yields sequence of fibonaci numbers.


# a most simple co-routine that yields all integers in range: from .. to

sub myrange(  to )

   a = 1
   b = 1
   
   i = 1
   while i <= to

     # yield function puts this thread to sleep and will wake up the caling thread
     # it also returns a value to the calling thread.

     nextval = a + b

     threadyield0( ~yieldval nextval )

     a = b
     b = nextval
     i = i + 1
   end
end

#
# Invoke the co-routine via the API.
#

th := makethread( ~func myrange )

th( ~to 10)

[ running, value ] = threadyieldvalue( ~thread th )

while( running != 0 )

  println( ~msg 'thread returned [ value ]' )

  [ running, value ] = resumethread( ~thread th )
end

stopthread( ~thread th )

