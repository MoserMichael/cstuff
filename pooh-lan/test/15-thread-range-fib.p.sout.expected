thread returned 2
thread returned 3
thread returned 5
thread returned 8
thread returned 13
thread returned 21
thread returned 34
thread returned 55
thread returned 89
thread returned 144
